### DISCUSSION TITLE
Choosing a date — we want to know your use cases

### DISCUSSION THREAD

**Discussion by christopherthomasdesign** [alphagov/govuk-design-system#2375] [Prioritised components and patterns]
We want to find out if adding a 'date picker' component to the Design System is a good idea. If not, we'd also consider adding more guidance on how to build one in the 'Dates' pattern.

## About this work

This proposed component will help users select a date using a calendar. 'Date pickers' are usually used to help users choose a date in the future, such as to make an appointment.

This work might also include improvements to:
* [Dates](https://design-system.service.gov.uk/patterns/dates)
* [Date input](https://design-system.service.gov.uk/components/date-input/)

## What we're looking for

We're currently looking for:
- examples of date pickers in services: screenshots, prototypes, links to live services
- use cases: explanations of why a 'date picker' was used in a service instead of a 'date input', or something else
- research: how well 'date pickers' tested with users to complete different tasks

This will help us understand the scenarios where 'date pickers' are used. We'd also like to understand how much service teams and users would benefit from a standardised 'date picker' in the Design System.

## Existing examples

[Digital Scotland](https://designsystem.gov.scot/components/date-picker/)

**Comment by davidjohnson444** [comment]
Our service allows users to claim compensation for injuries and illness due to Armed Forces service.  We ask for details of when people served, dates of incidents and hospital treatment, sometimes from many years ago.  It can be difficult for users to remember exact details.

Research with our processing teams told us it's hugely helpful to have at least an approximate year when trying to track down evidence to request on a claim (hospital notes etc).  However, users told us that the standard date pattern made them assume "_I have to get this right or it'll error me_".  We therefore decided to validate to just a year, guide users to enter an approximate year if they were unsure and introduced a checkbox "This date is approximate" in place of "I do not know".    Almost instantly this reassured our customer users in moderated testing and we received strong evidence the approach was having the desired effect over time, i.e. users telling us roughly when, for example they received treatment, whilst feeling reassured they were 'allowed' to not be exact.  

As a result, I'm not sure 'choose a date' would have worked for us because it's so specific.  I thought I'd share this in case others had similar considerations.
  
![Approx Dates](https://user-images.githubusercontent.com/78865338/196179797-9feaa408-0cfb-46ef-9c92-d74dfa43aed0.png)

**Comment by jessdenney** [comment]
HMRC Online Payments Service

As part of researching future dated payments with users we have been testing a date picker. 
Our thoughts were if its not a known date/event like a birthday it might help users plan for timing payments. For example, helping cash flow by timing it after payday, or even to help not schedule it on a weekend.

We've done 2 rounds of research so far, its shown to be quite well received and users reflected our thoughts- although one user noted that mon/sun can alternate as the first day of the week which can be confusing.

I did look at [Digital Scotland](https://designsystem.gov.scot/components/date-picker/) to see what guidance/format they had gone with before but couldn't make it match very well. I also looked at examples on [CJS](https://cjscpp-design-patterns.herokuapp.com/design_patterns/form-elements-date-picker) and [HMCTS](https://hmcts-design-system.herokuapp.com/components/calendar-date-picker)


![Screenshot 2022-10-17 at 16 49 36](https://user-images.githubusercontent.com/73240363/196223841-6722e62c-2b2f-4bad-abbc-dc5ded2431cb.png)

**Reply by christopherthomasdesign** [reply]
Thanks for sharing @jessdenney ! And for the other examples from CJS and HMCTS.

Have you tested with assistive tech users at all yet?

**Reply by jessdenney** [reply]
Hi no we haven't yet all testing has also been on desktop as well so might need more evidence before we can prove a really strong case for it. We do think it might be beneficial to neurodivergent users because of the visual element and helping to plan ahead.

We are planning to reach out to the accessibility team to get some insight from them about considerations we may need to think about in terms of WCAG.

**Reply by christopherthomasdesign** [reply]
Sounds good, let us know if you learn anything more 👍

**Reply by jessdenney** [reply]
UPDATE: 
I've managed to get date picker in our prototype that looks and functions a lot more similar to the driving theory test booking service. I was advised by accessibility that this would be the best route to take.

I would still like it to be closer to the pattern used on Digital Scotland as it looks more in keeping with GDS and is more accessible.
Is there any steer or guidance at implementing up and coming components like this as its really difficult to do if you don't have a dev background?

![Screenshot 2022-12-14 at 16 22 05](https://user-images.githubusercontent.com/73240363/207650952-4ff18ccf-f09f-4a4d-a333-df2a8cfa216f.png) 
[link to prototype](https://open-banking-prototype.herokuapp.com/4-3/pay/when-do-you-want-to-pay) (usual creds)
<img width="1242" alt="Screenshot 2022-12-14 at 16 23 11" src="https://user-images.githubusercontent.com/73240363/207651231-488eadb1-ee03-45e5-bf25-2c9167da6ffb.png">

**Reply by jessdenney** [reply]
Update: 
This is the final design we are going ahead with for the date picker.

In summary...

Context: Scheduling a future payment to HMRC using open banking

User need: as the date entered is likely to not be a 'known date' having a date picker helps the user plan. The calendar view allows the dates to be seen in context . The user can avoid scheduling a payment on a Saturday or Sunday for example when payments may take longer to process, or can plan around incoming and out going payments to manage cash flow.

We have decided to go with a version similar to the [Digital Scotland](https://designsystem.gov.scot/components/date-picker/) one as this one is the most accessible I have been able to find, it fits in well with the style of GDS and you can select a date range. 

We are making a few changes:
- Adding the button for the date picker on to the standard date input component to be consistent with GDS and more accessible
- Changing the styling of the buttons from Scottish blue to Gov green
- Adding in additional labels to make it more accessible

We are interested to monitor this to see how users interact with the page

This is a screenshot of our design
<img width="415" alt="Screenshot 2023-02-23 at 16 20 11" src="https://user-images.githubusercontent.com/73240363/220975898-d9c48ac4-6bc7-40be-983f-fedc2dab33fc.png">

**Reply by wyllb** [reply]
Hi, wanted to put a note that the above is implemented and integrates with the existing GDS `DateInput` component. If the browser isn't running JavaScript then it isn't visible. A limit can be indicated for a selectable date range. Full accessibility testing hasn't been completed yet, but appropriate aria-labels are in place for screen readers and it's keyboard accessible.

<img width="691" alt="image" src="https://user-images.githubusercontent.com/125358199/232029585-f6e47555-c01b-428b-960d-8a862ad3baf4.png">

**Reply by christopherthomasdesign** [reply]
Thanks for sharing this!

@wyllb just to clarify, when you say it isn't available if the browsers not running JS, do you mean that just the calendar button and date picker don't show? So it falls back to just the standard date inputs?

**Reply by wyllb** [reply]
Yep, that's correct. It just falls back to the date input fields. The button and dialog are completely absent from the DOM in this case.

**Reply by christopherthomasdesign** [reply]
Makes sense, quite neat! Thanks again 👍

**Reply by ghost** [reply]
@wyllb I'm building a UI for a HO project. I need a date picker, and yours seems perfect. Would it be possible to copy, and how could I get access to it? Thanks

**Reply by jessdenney** [reply]
Hi @AllenConquest-6p6 you can see it in our prototype [here](https://open-banking-prototype.herokuapp.com/4-3/pay/when-do-you-want-to-pay) (UN:prototypes PW:P4yment5). Or in Pay-frontend github [here](https://github.com/hmrc/pay-frontend/blob/8cc4fbd48219a0897ef45968285b928cf6b0df01/app/assets/javascripts/datepicker.js#L5) and [here](https://github.com/hmrc/pay-frontend/blob/main/app/assets/stylesheets/datepicker.scss).
Happy to chat on X-gov slack if I can help further

**Comment by pbrocketmoj** [comment]
MOJ Visit Someone in Prison

We're currently iterating a date/time picker that allows users (call centre agents) to select a date and see available time slots to book a prison visit on behalf of a visitor over the phone. It's quite a large 'date picker' as it also needs to display time slots and prisoner schedule alongside the dates.

Our initial design was accordion based with available time slots shown when the accordion was open. This tested really well, but after going live we got feedback that users would prefer to see dates in context as often people would be asking to visit e.g. 'next Tuesday'.

![6 Time and date - MVP+1](https://user-images.githubusercontent.com/85356666/210397462-5da26670-4bd7-4b41-9dbe-6c4438ead29e.png)

What we have been testing is a calendar style component, which is divided into 28 day blocks as this is the timeframe in which visits can be booked. There are several different states for a date which are shown on the calendar - a date could either be completely unavailable as there are no visits that day, there could be visits on that date but they are fully booked/unavailable for that prisoner, or there could be visits on a date that are able to be booked. The fourth state would be the 'selected' date, and once a user clicks beyond 28 days they see another state which is for future visits accompanied by the date they are available from.

![Desktop calendar - basic](https://user-images.githubusercontent.com/85356666/210398002-28d55679-6543-4481-8e64-437b7960261a.png)
![Desktop calendar - future](https://user-images.githubusercontent.com/85356666/210398006-bff5de4a-0eee-4509-bec5-89b620cf6920.png)

We've been testing this with current users but have only got as far as Figma prototypes currently, we're looking at how to make this fully accessible but have not done any accessibility testing yet. 

The other part of this is making something that allows visitors to book for themselves online, and there is a calendar component being used for this on the existing prison visit booking service (which I don't have access to) - we'll be replacing this, but as most users are on mobile we'll be looking at other ways of doing this that aren't calendar based as it doesn't scale down to a really small mobile size very well.

**Reply by pbrocketmoj** [reply]
To update on this, we did some more iterating and research on the calendar component to get to the following designs:

![Desktop calendar - basic date selected](https://github.com/alphagov/govuk-design-system/assets/85356666/d49d15a4-d748-4594-b442-cf37cf9dbec7)
![Desktop calendar - basic future](https://github.com/alphagov/govuk-design-system/assets/85356666/d80b6a56-8b84-4649-b81d-4f26c66b5804)
![Desktop calendar - basic](https://github.com/alphagov/govuk-design-system/assets/85356666/44e4c83b-8572-4d0b-ab68-fc8cd6002bec)

These tested well with users, minimising any reliance on colour to convey information. These haven't been developed yet as we've had other high priority tickets, and from testing most users are actually very happy with the accordion-style date picker.

**Reply by pbrocketmoj** [reply]
Latest update on this for anyone who is interested:

We ended up developing a calendar date picker for the public service of book a prison visit
![Desktop calendar MVP](https://github.com/user-attachments/assets/2c563f7c-261a-4ebc-998f-92ca989b07c7)

https://www.figma.com/design/qgKJI2rYX3kC8l2T9JRPKf/Calendar-Design?node-id=1604-483

This has just gone through an accessibility audit and passed with flying colours:

"The date selection component is implemented very well, both visually and in terms of screen reader accessibility. Screen reader users are clearly informed about the date they are focusing on, the corresponding day of the week, and the number of available visit slots as they tab through the calendar.

Upon selecting a date, focus automatically shifts to the available slots, allowing users to proceed directly to time booking without having to tab through all remaining dates. Additionally, after selecting a date, users are presented with an option to choose a different date via the ‘Choose a different date above’ link. This link takes users back to the top of the calendar datepicker, making it easier to change their date selection without needing to navigate backwards through all the dates.

This design is very user-friendly, particularly for screen reader users, as it streamlines the process of selecting a date and time."
